### DISCUSSION TITLE
Make GitHub action use PATs instead of Repository secrets

### DISCUSSION THREAD

**Discussion by AnthonyMonterrosa** [community/community#25001] [Actions]
<p>Hello,</p>
<p>I am having difficulty creating an action to login to Github’s Container Registry.</p>
<p>In the <a href="https://github.com/docker/login-action" rel="noopener nofollow ugc">README for the docker/login-action repository</a>, it says to login to <a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a> (the container registry URL), you must use this</p>
<pre><code class="lang-auto">jobs:
  login:
    runs-on: ubuntu-latest
    steps:
      -
        name: Login to GitHub Container Registry
        uses: docker/login-action@v1
        with:
          registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.CR_PAT }}
</code></pre>
<p>where CR_PAT is a PAT with package read, write, and delete access.</p>
<p>I have created the CR_PAT with correct permissions, along with an ALL that I also used to see if it was a permissions issue</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/e/5/e5fe2e849dd9b84754e65a85ea2bd26d2bac24fd.png" data-download-href="/uploads/short-url/wOBVcerQ9k4QOE5ougi05DmogEd.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/12419195/181088084-3d687cdd-905d-4a89-b170-2ffdd0773726.png" alt="image" data-base62-sha1="wOBVcerQ9k4QOE5ougi05DmogEd" width="690" height="130" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/e/5/e5fe2e849dd9b84754e65a85ea2bd26d2bac24fd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">727×137 15.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>But, my login action would fail with the error below.</p>
<pre><code class="lang-auto">Run docker/login-action@v1
  with:
    registry: ghcr.io
    username: AnthonyMonterrosa
    logout: true
Error: Username and password required
</code></pre>
<p>After some testing, I realized that my PAT was not being found from the <code>${{ secrets.CR_PAT }}</code> expression, and confirmed when I saw both PATs marked as <code>Never used</code>.</p>
<p>Further, I realized that <code>${{ secrets.CR_PAT }}</code> could be looking in the repository secrets list, so I made a <code>CR_PAT_REPOSITORY</code> secret with a value equal to my <code>CR_PAT</code>'s access token.</p>
<p>and used it in an updated action .yml</p>
<pre><code class="lang-auto">        uses: docker/login-action@v1
        with:
          registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.CR_PAT_REPOSITORY }}
</code></pre>
<p>and this returned with a filled-in <code>password</code> field, but was denied access (which I kind of expected. How can a repository use a PAT?).</p>
<pre><code class="lang-auto">Run docker/login-action@v1
  with:
    registry: ghcr.io
    username: AnthonyMonterrosa
    password: ***
    logout: true
🔑 Logging into ghcr.io...
Error: Error response from daemon: Get https://ghcr.io/v2/: denied: denied
</code></pre>
<p>So, I know that <code>${{ secrets.X }}</code> looks in the repository’s secret list for a value, but does not search in my account’s PATs. Is there a way to make it look there? I’ve just been following the instructions in GitHub documentation and the <code>docker/login-action</code> repository README, so at this point I’m pretty confused.</p>
<p>For the record, I have enabled <code>Improved container support</code> for GitHub’s container registry.</p>
<p>Future thanks.</p>

**Comment by AnthonyMonterrosa** [comment]
<p>Because I’m a new user, I wasn’t able to put in all the pictures and links that I wanted into the post. Please feel free to ask if you want to have more information.</p>

**Comment by mskyttner** [comment]
<p>I run into this too. An earlier setup that I had before 2021 worked with settings like that. The <a href="https://github.com/mskyttner/duckdb-cli-builds/runs/1442603414?check_suite_focus=true" rel="noopener nofollow ugc">logs from there</a> shows a “password” line with a masked password printed (***). I wonder if someething (note misspelling because the spellcheck will not allow me to write meeth without doubling the e) has changed since then. When I used a similar gha today I saw what you are seeing, approximately. I get  “Error: Username and password required”, and even after trying to generate a new CR_PAT and add it specifically to the repo settings (Environment &gt; Secrets) it doesn’t get picked up. I was expecting that ${{ secrets.CR_PAT}} would be picked up but it doesn’t seem so.</p>
<p>Any solution yet?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="AnthonyMonterrosa" data-post="1" data-topic="155191">
<div class="title">
<div class="quote-controls"></div>
 AnthonyMonterrosa:</div>
<blockquote>
<p>So, I know that <code>${{ secrets.X }}</code> looks in the repository’s secret list for a value, but does not search in my account’s PATs. Is there a way to make it look there?</p>
</blockquote>
</aside>
<p>No, because that’d be a massive security hole. You have to explicitly provide secrets to actions where you want to use them (see <a href="https://docs.github.com/en/actions/reference/encrypted-secrets#accessing-your-secrets">Accessing your secrets</a>).</p>
<p>As far as the login problem is concerned, your token seems to be lacking the <em>read:packages</em> permission.</p>

**Comment by mskyttner** [comment]
<p>I checked my CR_PAT (aga_in), both the one I used successfully before and the new one I attempted to use, both have the <em>read:packages</em> permission.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/8/4/84b1e688b20eb549521a8d10f44c402129369564.png" data-download-href="/uploads/short-url/iVSaVmoQtBGK8EI4jco0YOLVJ3e.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1715840/181088088-066390dc-10f3-4582-956b-72d82b4b4ff6.png" alt="image" data-base62-sha1="iVSaVmoQtBGK8EI4jco0YOLVJ3e" width="690" height="430" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/84b1e688b20eb549521a8d10f44c402129369564_2_690x430.png, https://aws1.discourse-cdn.com/github/original/3X/8/4/84b1e688b20eb549521a8d10f44c402129369564.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/4/84b1e688b20eb549521a8d10f44c402129369564.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/84b1e688b20eb549521a8d10f44c402129369564_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">960×599 59.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Can it be some_thing else coming into play here?</p>
<p>(Using underscores to circu_mvent the censoring of potentially upsetting strings inside words)</p>

**Comment by AnthonyMonterrosa** [comment, marked as answer]
<p>My solution was to create a PAT with read/write/delete package permissions, and to copy the generated token value to a repository secret, and then to reference that repository secret in the action. That seems to be the only way to use PATs in actions, by copying them to the action’s repo.</p>

**Comment by mskyttner** [comment]
<p>That works well for me to, it needs to be a repository secret not an environment secret. Maybe you should put it as an accepted answer, at least it solved the problem for me.</p>
